dfinity/nns-dapp
# NNS1-2905: Stop encoding the account map in accounts store

# Motivation

We want to remove the field `default_account_transactions` from the `Accounts` type.
This is easier if the `Accounts` type is encoded in fewer places.
One place where we encode the `Accounts` type, that is not necessary, is when we encode the `AccountsStore` we encode an empty `BTreeMap` with `Account` values.
This used to be where we store the accounts, but the accounts are now stored in stable structures so now we just encode an empty `BTreeMap` as placeholder.
The value type of the empty map does not matter as long as it can be decoded in `post_upgrade`.
The `candid::Empty` type can be decoded to any type so by replacing `Account` with `candid::Empty` as the value type of this `BTreeMap` we have more freedom with changing the structure of `Account` in future PRs.

While this PR does some of the cleanup for the stable structures migration, it does not aim to remove any more than necessary to stop encoding/decoding `Account` in `AccountsStore::encode`/`::decode`.

# Changes

1. In `AccountsStore::encode` always encode an empty accounts map instead of checking if the `accounts_db` has a non-empty map. The stable structures migration has finished so we no longer have the DB in a map.
2. In `AccountsStore::encode` change the type of the encoded empty map to have value type `candid::Empty`.
3. In `AccountsStore::decode` decode the accounts map into type `candid::Reserved` which can decode anything, and then ignore it.
4. In `AccountsStore::decode` don't try to fall back on stats from the map because we always have stats from the DB.
5. In `AccountsStore::decode` initialize a default accounts DB because it will be replaced anyway.

# Tests

1. In `state_heap_contents_can_be_serialized_and_deserialized`, drop the accounts map before comparing. We no longer encode the accounts map because we expect the accounts DB to take care of that.
2. Remove `test-upgrade-map` and `test-upgrade-map-stable` migration tests which require the accounts map.
3. Remove the `migration_happy_path` and `interrupted_migration` PocketIC tests.

# Todos
